corehq/apps/receiverwrapper/util.py

```diff
@@ -209,17 +211,23 @@ def get_app_and_build_ids(domain, build_or_app_id):
     return build_or_app_id, None
 
 
-def from_demo_user(form_json):
+def from_demo_user(form_json, domain):
     """
     Whether the form is submitted by demo_user
     """
     from corehq.apps.users.util import DEMO_USER_ID
     try:
-        # require new-style meta/userID (reject Meta/chw_id)
-        if form_json['meta']['userID'] == DEMO_USER_ID:
-            return True
+        user_id = form_json['meta']['userID']
     except (KeyError, ValueError):
         return False
+    else:
+        # require new-style meta/userID (reject Meta/chw_id)
+        if user_id == DEMO_USER_ID:
+            return True
+        if settings.SERVER_ENVIRONMENT in settings.ICDS_ENVS:
+            user = CommCareUser.get_by_user_id(user_id, domain)
```

Comment:
You can get the user from the request to avoid an extra cache hit: `request.couch_user`

Reply:
https://github.com/dimagi/commcare-hq/pull/25615/commits/0f7e313d07d817814e4fc19c6478e6e1687ea781